File: libs/statemanager/src/main/kotlin/mono/state/command/mouse/LineInteractionMouseCommand.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.state.command.mouse

import mono.common.MouseCursor
import mono.graphics.geo.DirectedPoint
import mono.graphics.geo.MousePointer
import mono.graphics.geo.Point
import mono.shape.command.MoveLineAnchor
import mono.shape.command.MoveLineEdge
import mono.shape.selection.SelectedShapeManager
import mono.shape.shape.Line
import mono.shapebound.LineInteractionPoint
import mono.state.command.CommandEnvironment

/**
 * A [MouseCommand] for moving anchor points or edges of Line shape.
 */
internal class LineInteractionMouseCommand(
    private val lineShape: Line,
    private val interactionPoint: LineInteractionPoint
) : MouseCommand {
    override val mouseCursor: MouseCursor? = when (interactionPoint) {
        is LineInteractionPoint.Anchor -> MouseCursor.CROSSHAIR
        is LineInteractionPoint.Edge -> null
    }

    private val hoverShapeManager = HoverShapeManager.forLineConnectHover()

    override fun execute(
        environment: CommandEnvironment,
        mousePointer: MousePointer
    ): MouseCommand.CommandResultType {
        when (mousePointer) {
            is MousePointer.Drag -> move(
                environment,
                mousePointer.boardCoordinate,
                isUpdateConfirmed = false,
                justMoveAnchor = !mousePointer.isWithShiftKey
            )

            is MousePointer.Up -> move(
                environment,
                mousePointer.boardCoordinate,
                isUpdateConfirmed = true,
                justMoveAnchor = !mousePointer.isWithShiftKey
            )

            is MousePointer.Down,
            is MousePointer.Click,
            is MousePointer.DoubleClick,
            is MousePointer.Move,
            MousePointer.Idle -> Unit
        }

        return if (mousePointer == MousePointer.Idle) {
            MouseCommand.CommandResultType.DONE
        } else {
            MouseCommand.CommandResultType.WORKING
        }
    }

    private fun move(
        environment: CommandEnvironment,
        point: Point,
        isUpdateConfirmed: Boolean,
        justMoveAnchor: Boolean
    ) {
        when (interactionPoint) {
            is LineInteractionPoint.Anchor ->
                moveAnchor(environment, interactionPoint, point, isUpdateConfirmed, justMoveAnchor)

            is LineInteractionPoint.Edge ->
                moveEdge(environment, interactionPoint, point, isUpdateConfirmed)
        }
    }

    private fun moveAnchor(
        environment: CommandEnvironment,
        interactionPoint: LineInteractionPoint.Anchor,
        point: Point,
        isUpdateConfirmed: Boolean,
        justMoveAnchor: Boolean
    ) {
        val edgeDirection = environment.getEdgeDirection(point)
        val direction =
            edgeDirection?.normalizedDirection ?: lineShape.getDirection(interactionPoint.anchor)
        val anchorPointUpdate = Line.AnchorPointUpdate(
            interactionPoint.anchor,
            DirectedPoint(direction, point)
        )
        val connectShape = hoverShapeManager.getHoverShape(environment, anchorPointUpdate.point)
        environment.setFocusingShape(
            connectShape.takeIf { !isUpdateConfirmed },
            SelectedShapeManager.ShapeFocusType.LINE_CONNECTING
        )
        environment.shapeManager.execute(
            MoveLineAnchor(
                lineShape,
                anchorPointUpdate,
                isUpdateConfirmed,
                justMoveAnchor = justMoveAnchor,
                connectShape = connectShape
            )
        )
        environment.updateInteractionBounds()
    }

    private fun moveEdge(
        environment: CommandEnvironment,
        interactionPoint: LineInteractionPoint.Edge,
        point: Point,
        isUpdateConfirmed: Boolean
    ) {
        environment.shapeManager.execute(
            MoveLineEdge(
                lineShape,
                interactionPoint.edgeId,
                point,
                isUpdateConfirmed
            )
        )
        environment.updateInteractionBounds()
    }
}

File: monosketch-svelte/src/lib/mono/state-manager/command/mouse/line-interaction-mouse-command.ts
import { DirectedPoint, Direction, type Point } from "$libs/graphics-geo/point";
import { LineInteractionPoint } from "$mono/shape-interaction-bound/interaction-point";
import { MoveLineAnchor, MoveLineEdge } from "$mono/shape/command/line-commands";
import { ShapeFocusType } from "$mono/shape/selected-shape-manager";
import type { Line } from "$mono/shape/shape/line";
import type { CommandEnvironment } from "$mono/state-manager/command-environment";
import { HoverShapeManager } from "$mono/state-manager/command/mouse/hover-shape-manager";
import { MouseCommand } from "$mono/state-manager/command/mouse/mouse-command";
import { MouseCursor } from "$mono/workspace/mouse/cursor-type";
import { type MousePointer, MousePointerType } from "$mono/workspace/mouse/mouse-pointer";
import CommandResultType = MouseCommand.CommandResultType;

/**
 * A [MouseCommand] for adding a new Line shape.
 */
export class LineInteractionMouseCommand implements MouseCommand {
    mouseCursor?: MouseCursor;
    private readonly hoverShapeManager: HoverShapeManager = HoverShapeManager.forLineConnectHover();


    constructor(private shape: Line, private interactionPoint: LineInteractionPoint.Base) {
        this.mouseCursor = interactionPoint instanceof LineInteractionPoint.Anchor
            ? MouseCursor.CROSSHAIR
            : undefined;
    }

    execute(environment: CommandEnvironment, mousePointer: MousePointer): MouseCommand.CommandResultType {
        switch (mousePointer.type) {
            case MousePointerType.DRAG:
                this.move(environment, mousePointer.boardCoordinate, false, !mousePointer.isWithShiftKey);
                break
            case MousePointerType.UP:
                this.move(environment, mousePointer.boardCoordinate, true, !mousePointer.isWithShiftKey);
                break
        }

        return mousePointer.type === MousePointerType.IDLE ? CommandResultType.DONE : CommandResultType.WORKING;
    }

    private move(
        environment: CommandEnvironment,
        point: Point,
        isUpdateConfirmed: boolean,
        justMoveAnchor: boolean,
    ): void {
        if (this.interactionPoint instanceof LineInteractionPoint.Anchor) {
            this.moveAnchor(
                environment,
                this.interactionPoint,
                point,
                isUpdateConfirmed,
                justMoveAnchor,
            );
        } else if (this.interactionPoint instanceof LineInteractionPoint.Edge) {
            this.moveEdge(
                environment,
                this.interactionPoint,
                point,
                isUpdateConfirmed,
            );
        }
    }

    private moveAnchor(
        environment: CommandEnvironment,
        interactionPoint: LineInteractionPoint.Anchor,
        point: Point,
        isUpdateConfirmed: boolean,
        justMoveAnchor: boolean,
    ): void {
        const edgeDirection = environment.getEdgeDirection(point);
        const direction = edgeDirection ? Direction.normalizedDirection(edgeDirection) :
            this.shape.getDirection(interactionPoint.anchor);

        const anchorPointUpdate = {
            anchor: interactionPoint.anchor,
            point: new DirectedPoint(direction, point),
        };

        const connectShape = this.hoverShapeManager.getHoverShape(
            environment,
            anchorPointUpdate.point.point,
        );

        environment.setFocusingShape(
            isUpdateConfirmed ? null : connectShape,
            ShapeFocusType.LINE_CONNECTING,
        );

        environment.shapeManager.execute(
            new MoveLineAnchor(
                this.shape,
                anchorPointUpdate,
                isUpdateConfirmed,
                justMoveAnchor,
                connectShape,
            ),
        );

        environment.updateInteractionBounds();
    }

    private moveEdge(
        environment: CommandEnvironment,
        interactionPoint: LineInteractionPoint.Edge,
        point: Point,
        isUpdateConfirmed: boolean,
    ): void {
        environment.shapeManager.execute(
            new MoveLineEdge(
                this.shape,
                interactionPoint.edgeId,
                point,
                isUpdateConfirmed,
            ),
        );

        environment.updateInteractionBounds();
    }
}
